### PR TITLE
[inductor][ROCm] Flush the LLC, not just L2, when autotuning

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #192159

Autotuning times each candidate with a cache flush between iterations:
`benchmark_gpu` zeroes a buffer sized by `get_device_cache_size()` so the input
is evicted and every iteration measures a cold kernel. Sizing that buffer to the
device's exact reported cache rather than Triton's flat 256 MB was deliberate
(https://github.com/pytorch/pytorch/pull/133058) and is the right principle, but
it assumes the reported size is the last-level cache. On MI300X it is not: the
hierarchy is a 4 MB L2 per XCD, then a 256 MB last-level cache in front of HBM,
and hipDeviceProp exposes only the per-XCD L2. So the flush comes out 64x too
small, zeroing 4 MB evicts nothing from a 256 MB cache, and every candidate is
timed reading out of cache instead of HBM -- exactly the regime where
memory-bound configs are indistinguishable.

So autotune ranks by noise. For layer_norm fwd fp32 M=4096 N=4096, configs
(XBLOCK=1, R0_BLOCK=4096, warps=4), (1, 2048, 2) and (1, 2048, 8) score
50.7 / 51.7 / 51.1 us against true times of 90.9 / 54.7 / 64.5 us: a 1.64x
spread collapses to 2.6% and the worst config ranks first.

The fix is to flush 256 MB on ROCm instead of the 4 MB the device reports. New
`get_cache_flush_size()` owns that rule -- on ROCm the larger of Triton's 256 MB
and whatever the device reports, elsewhere exactly the device cache size -- and
both `benchmark_gpu` implementations call it, replacing the inline ROCm case
`TorchProfilerBenchmarker` already carried
(https://github.com/pytorch/pytorch/pull/175097). `get_device_cache_size()` is
no longer overloaded to mean "flush size" and returns None when the device
reports no cache. Non-ROCm is unchanged by construction. The gate is arch-wide
because no arch exposes a queryable LLC size and Triton's AMD backend flushes
256 MB on every one, so matching it keeps all three benchmarkers ranking alike.

This costs ROCm compile time: one `zero_()` goes 4.2 -> 55.7 us and
`benchmark_gpu` zeroes 110-210 times per candidate. Five bf16 matmuls under
max-autotune-no-cudagraphs, 6 interleaved separate-process rounds per arm:
6.89 s base against 7.77 s, +12.8% over 74 benchmarked configs, disjoint
distributions.

tritonbench on a perf_determinism-locked MI300X, median of interleaved
separate-process rounds per arm with a fresh inductor cache each. layer_norm fwd
fp32 M=4096, 30 shapes, 5 rounds per arm:

    N       base us   new us   ratio
    1536       80.8     26.0    3.11
    2048       83.8     28.0    2.99
    2560       83.6     36.7    2.28
    3072       83.6     39.3    2.13
    3584       90.7     43.7    2.07
    4608      100.1     64.8    1.54

Geomean 1.221 over all 30, and 1.006 over the 19 at N>=6656 where the input no
longer fits the LLC. Selection also stops being bimodal: base spreads >10%
run-to-run on 9 of 30 shapes against 3 here. Counters put traffic within 6%
across configs whose latency spans 3.1x, so the lever is achieved bandwidth, not
bytes moved. Scope checks: rms_norm +1.6% geomean over 6 shapes (H=4096 +16.5%,
H=8192 +13.5%), softmax fp16 neutral at 1.003 over 98.

Regressions, each deterministic per shape:

    op          shape      delta   cause
    rms_norm    H=1024    -11.9%   selection swing between configs 15% apart
    rms_norm    H=2048     -5.5%   genuinely slower selected config
    layer_norm  N=5120     -3.6%   base bimodal, this arm deterministic at 74.3
    softmax     scattered   +-8%   no direction, coordinate-descent search paths

Both rms_norm shapes are sub-20 us kernels whose 16 MB footprint sits far inside
the LLC, the regime where an LLC-sized flush most changes what "cold" means. All
ratios are raw; the eager control itself moves up to 5% at the smallest shapes,
so sub-1% readings are not worth interpreting.

Follow-up: `return_mode="min"` scores by the fastest of ~100 timed iterations,
which under this flush is a first-iteration transient. The 100 warmup passes do
not remove that transient -- they make it reach every candidate, which is what
keeps the comparison fair; cutting them to 10 denied it to whichever candidate
runs first and mis-ranked. Scoring with the median would be position-independent
by construction, allow a much shorter warmup, recover most of the +12.8%, and
align with `TritonBenchmarker`, which already uses median. It changes semantics
for every backend including CUDA, so it needs separate validation.

Authored with the assistance of an AI coding agent.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

**Comment by desertfire** [comment]
https://hud.pytorch.org/benchmark/compilers_regression?renderGroupId=main&time.start=2026-07-30T00%3A00%3A00.000Z&time.end=2026-08-06T23%3A59%3A59.999Z&filters.repo=pytorch%2Fpytorch&filters.benchmarkName=compiler&filters.backend=&filters.mode=inference&filters.dtype=bfloat16&filters.deviceName=rocm+%28mi355x%29&filters.device=rocm&filters.arch=AMD+Radeon+Graphics&lbranch=gh%2Fdesertfire%2F752%2Fbase&rbranch=gh%2Fdesertfire%2F752%2Fhead&rcommit.commit=1f9d52fa1eb7543fa396915804770a859fe9618f&rcommit.workflow_id=30967695309&rcommit.date=2026-08-05T03%3A00%3A00Z&rcommit.branch=gh%2Fdesertfire%2F752%2Fhead&lcommit.commit=2d267070b5e7e0feb78ccfff29e7f5b6156cc0ca&lcommit.workflow_id=31018024122&lcommit.date=2026-08-05T16%3A00%3A00Z&lcommit.branch=gh%2Fdesertfire%2F752%2Fbase&maxSampling=110

The dashboard run shows it did improve the performance of torchbench inference (smaller models), but with larger memory peak size. Interestingly, it didn't move the needle on the training side.
